### DISCUSSION TITLE
Having issues with generate_taxa_test_long

### DISCUSSION THREAD

**Discussion by adahal123** [cafferychen777/MicrobiomeStat#44] [Q&A] [answered]
####FEATURE LEVEL ANALYSIS#######

**This code works great**
test.list <- generate_taxa_trend_test_long(
  data.obj = ps.obj,
  subject.var = "pid",
  time.var = "visit",
  group.var = "race",
  adj.vars = "sex",
  prev.filter = 0.0,
  abund.filter = 0.0,
  feature.level = c("Genus"),
  feature.dat.type = c("proportion")
)

generate_taxa_trend_volcano_long(
  data.obj = ps.obj,
  group.var = "race",
  time.var = "visit",
  test.list = test.list,
  feature.sig.level = 0.1,
  feature.mt.method = "none"
)

**This code below does not work:** 
result2 <- generate_taxa_test_long(
  data.obj = ps.obj,
  subject.var = "pid",
  time.var = "visit",
  group.var = "race",
  adj.vars = "sex",
  prev.filter = 0.0,
  abund.filter = 0.0,
  feature.level = c("Genus"),
  feature.dat.type = c("proportion")
)

Results:
<img width="873" alt="Screenshot 2024-04-05 at 2 32 50 PM" src="https://github.com/cafferychen777/MicrobiomeStat/assets/31011373/037d7b3d-9490-4381-bd5c-90a1155a607e">

It then excludes samples from visit 2 and visit 3 and then shows this message: 

<img width="844" alt="Screenshot 2024-04-05 at 2 32 56 PM" src="https://github.com/cafferychen777/MicrobiomeStat/assets/31011373/0e796f98-f6c0-408c-9785-548822ef58b0">

As you can see, generate taxa trend works and so do the alpha diversity functions with these same variables specified. Sorry for asking so many questions.

**Comment by cafferychen777** [comment]
Hi @adahal123,

Thank you for bringing this issue to our attention. It appears that there might be a problem with the experimental design that is causing the generate_taxa_test_long function to exclude samples from visit 2 and visit 3.

To further investigate this, could you please try running your data object through the `mStat_summarize_data_obj` function with the appropriate arguments for `time.var` and `group.var`? For example:

```R
summary_result <- mStat_summarize_data_obj(data.obj, time.var = "visit", group.var = "race")
```

This will help us understand if the issue is indeed related to the experimental design or if there's something else we need to address.

Please let me know how it goes, and if you have any further questions, feel free to ask.

Best regards,
Caffery Yang

**Reply by adahal123** [reply]
The experimental setup works for most of the functions such as `generate_taxa_trend_test_long` and 
`generate_taxa_trend_volcano_long` but does not work with `generate_taxa_test_long`

This is the output of the `mStat_summarize_data_obj`
<img width="522" alt="Screenshot 2024-04-06 at 11 25 54 AM" src="https://github.com/cafferychen777/MicrobiomeStat/assets/31011373/18330cef-73c6-40f8-b934-fd0570e96265">

<img width="564" alt="Screenshot 2024-04-06 at 11 27 50 AM" src="https://github.com/cafferychen777/MicrobiomeStat/assets/31011373/cda2ac20-f9fb-4e5d-bd0a-e18573bc51cf">

**Reply by cafferychen777** [reply]
Hi adahal123,

Thank you for your continued engagement with the MicrobiomeStat package and for bringing this to my attention.

Regarding the exclusion of samples from visit 2 and visit 3 in the `generate_taxa_test_long` function, I realize that this behavior is actually expected. The function is designed to analyze each time point separately, so the exclusion of samples from certain visits is a normal part of the process.

However, the error message you encountered, "number of levels of each grouping factor must be < number of observations (problems: pid)," is indeed concerning. One potential solution could be to convert the `group.var` (in your case, "race") to a factor. This might resolve the issue. If it doesn't, I would appreciate it if you could send your data privately to my email at cafferychen7850@gmail.com. This will allow me to take a closer look at the problem and provide a more specific solution.

I understand that this issue can be frustrating, and I'm here to help you resolve it. Please feel free to reach out with any further questions or concerns.

Best regards,
Caffery Yang

**Reply by cafferychen777** [reply]
Hi adahal123,

Thank you for your detailed feedback and for bringing these issues to our attention. I appreciate your patience as we work through them.

Regarding the generate_alpha_test_long function, I see that you've used numeric values for t0.level and character values for ts.levels. This inconsistency might be causing some of the problems you're experiencing. Could you try using numeric values for both t0.level and ts.levels and see if that resolves the issue? For example:

```R
alpha_test <- generate_alpha_test_long(
  data.obj = ps.obj,
  alpha.obj = alpha.obj,
  alpha.name = c("shannon", "observed_species"),
  time.var = "visit",
  t0.level = 1,
  ts.levels = c(2, 3),
  group.var = "race",
  adj.vars = "sex"
)
```

Please let me know if this change resolves the problem or if you continue to encounter issues. We're committed to ensuring that the MicrobiomeStat package meets the needs of its users, so your feedback is invaluable.

Best regards,
Caffery Yang

**Reply by adahal123** [reply]
Ok this is for the alpha question which was in the other thread. 
Using numeric for t0 and ts does not work. As you can see from my previous examples in the other thread, putting numeric for t0 and character values for ts worked for other functions but just not for `generate_alpha_test_long`

Still getting the same error message as before.

**Reply by cafferychen777** [reply]
Hi, thanks for reaching out again! It seems like the issue you're encountering with the `generate_taxa_test_long` function might be related to how the time variable is being handled in different functions. In other functions, if the time variable is character or factor, it could cause problems, so I've added a step to convert it to numeric (`data.obj$meta.dat <- data.obj$meta.dat %>% dplyr::mutate(!!sym(time.var) := as.numeric(!!sym(time.var)))`). However, for `generate_alpha_test_long`, the test is conducted separately for each time point, so the format of the time variable shouldn't be a problem, and I didn't add this conversion step.

This might explain why it works for other functions but not for `generate_alpha_test_long`. I believe specifying `t0.level = 1` and `ts.levels = c(2,3)` should not cause issues in other functions, and I suggest doing the same in `generate_alpha_test_long` for consistency. Please let me know if this resolves the issue or if you have any further questions!

**Reply by adahal123** [reply]
Like I mentioned in my last post, specifying t0.level = 1 and ts.levels = c(2,3) does not work for `generate_alpha_test_long` does not work.

**Reply by cafferychen777** [reply, marked as answer]
Hi adahal123,

Thanks for reporting the issues you encountered with the `generate_taxa_test_long` and `generate_alpha_test_long` functions in the MicrobiomeStat package. 

Regarding the problem with `generate_taxa_test_long`, I understand that it's not working for you even though other related functions like `generate_taxa_trend_test_long` and `generate_taxa_trend_volcano_long` are working fine with the same dataset and parameters. However, I've been unable to reproduce this error using the example data. Without access to the specific dataset you are working with, it's difficult for me to investigate why this issue is occurring in your case.

Similarly, for the `generate_alpha_test_long` function, specifying numeric values for the `t0.level` and `ts.levels` arguments doesn't seem to be working based on the error message you shared. But again, I couldn't replicate this behavior with the example data I have.

If you are able to share a small subset of your data that demonstrates these problems (either publicly or privately), that would be extremely helpful. Having the actual data will allow me to dig deeper, reproduce the errors, and hopefully figure out what's causing these functions to fail in your specific use case.

Please let me know if you can provide some sample data, and I'll be happy to take a closer look. Apologies for the inconvenience, and thanks in advance for your help in debugging these issues.

Best regards,
Caffery
